### DISCUSSION TITLE
Parsing to dictionaries

### DISCUSSION THREAD

**Discussion by JaapWijnen** [pointfreeco/swift-parsing#248] [Q&A]
Hi! I've been doing some xml related parsing recently and I run into an issue when parsing attributes such as: 
`<xmlTag "attr1"="value1" "attr2"="value2" />`
When parsing these to a dictionary round tripping wil occasionally fail since dictionaries are unordered so the printed order will be reversed for example. 
So I switched to swift-collections' `OrderedDictionary` which fixed the problem. 
A few of swift-collections' types don't conform to collection however due to protocol requirement clashes with subscripts I believe. 

Specifically the `End()` parser requires Input to conform to Collection. Summarising I have two questions:
- Are there better ways to get around my ordered dictionary problem with xml/html attributes?
- Could we loosen the restrictions on the `End` parser to Sequence? 

Something like this seems to work in this case:
```swift
struct SequenceEnd<Input: Sequence>: Parser {
    func parse(_ input: inout Input) throws {
        var iterator = input.makeIterator()
        guard iterator.next() == nil else {
            throw SomeError()
        }
    }
}
```

Let me know what you think!

**Comment by stephencelis** [comment]
Hi @JaapWijnen! I converted this to a discussion since we use issues for bug tracking.

Can you share more of your use case? It sounds like you're parsing a substring/bytes _into_ an `OrderedDictionary`, so `OrderedDictionary` would be the _output_, not the _input_, right? Or are you trying to write a parser against an ordered dictionary?

**Reply by JaapWijnen** [reply]
Ah yes my bad, thanks! I have an XMLParser that parses into an ordered dictionary, thats correct. Im now creating an XSD parser that grabs the xml parsers output to further parse that down to an xsd format. Therefore there will be specific xsd tags that only expect specific attributes. So after parsing them id like to check that the ordered dictionary is empty. Thats my current usecase :) Does that make more sense?

**Reply by stephencelis** [reply]
Makes sense and should be possible with the tools we ship with! If you're having any particular problems, feel free to share what you have so far.

**Reply by JaapWijnen** [reply]
Yeah it's absolutely possible! I meant to ask would it be possible to drop End parser's Input restriction to Sequence instead of Collection to have better support for the swift-collection package for example. Or do you foresee other problems with that? 
(there could be other Parsers/ParserPrinters for which this might be interesting but I haven't run into them yet. 

Regarding that ordered dictionary. Would that be the way to go in order to make that roundtrippable? Or would you have a different suggestion? I'd rather use normal ones of course.

**Reply by stephencelis** [reply]
> Yeah it's absolutely possible! I meant to ask would it be possible to drop End parser's Input restriction to Sequence instead of Collection to have better support for the swift-collection package for example.

Oh! That sounds reasonable to me. Wanna try to PR and see if anything comes up?

**Reply by JaapWijnen** [reply]
Will do!

**Reply by JaapWijnen** [reply]
Done! #250
